### ISSUE TITLE
What should be happen when zfc.reallocate is called with 0 or 1 offerHandle?

### ISSUE THREAD

**Issue by DavidBruant** [Agoric/agoric-sdk#1017]
During the hackathon, i've seen the following pattern:
```js
zcf.reallocate(
  [offerHandle],
  [zcf.getCurrentAllocation(offerHandle)],
);
```

Although it's not wrong to do that, it's a no-op (the assets allocated with an offerHandle are re-allocated... to the same offerHandle)
it feels like this snippet is a mistake made by the author

I think that calling `zfc.reallocate` with 0 or 1 offerHandle should probably be considered a mistake and result at least in a `console.warn` or maybe an error thrown

**Comment by katelynsills** [comment]
Thanks so much for making this an issue!

I think we should throw. My reasoning is that calling with one offerHandle is always wrong (there is never a reason to do it) so we should give people that feedback immediately and at runtime (although linting beforehand is always good too). Being silent at runtime will allow something very mistaken to continue

**Comment by erights** [comment]
If I see in code `x += 0` it is likely that the code is either confused, or it is doing something subtle that I don't yet appreciate. For that, a lint rule makes sense. If it really is intentional, then an explicit `// eslint-disable-next-line` shows that, and hopefully provokes the developer into explaining why this is intentional.

If I see in code `x += d`, if `d` happens to be zero, I would find it unfortunate if `+=` rejected it with an error because "Why increment by `d` when `d` is zero? Likely means an error." If `+=` did indeed work that way, then I'd need to say something like `if (d !== 0) { x += d; }`.

These seem parallel, in that you get an error by default, but you can override it explicitly by code that, in the second case, still allows `d` to be zero.

But there's a big difference. Statically writing `+= 0` at least needs an explanation. Writing `+= d` where `d` might be zero should work by uniformity. There's no reason to presume it indicates the programmer was confused.

I understand that this example feels very different. But nevertheless, systems that work consistently in the zero case are easier to work with when feeding them computed (rather than static) values.

**Comment by erights** [comment]
Note that the pattern provoking this issue would be flagged by a static check.

**Comment by katelynsills** [comment]
I understand not wanting to have `x += d` throw if `d` is `0` so that the developer doesn't have to worry about that particular case. This makes sense because there are many situations in which you will add 0 - because the code is written such that `d` could be many things, including 0 and adding `d` as a rule makes sense. Thank you for the explanation. It was very helpful.

However, I don't think this is an analogous situation. We don't have any examples in which a developer will want to reallocate with a variable number of offers, which may be zero or 1.  Reallocations only make sense with at least two offers.

The situation that would be exactly analogous to your example would be reallocating with the same values as before. That we allow, as we should.

To use your example again, I think the comparable situation would be doing addition (which requires two or more terms) with only one term.

**Comment by erights** [comment]
If the reallocation is of a statically known set of offers, it'll be written with a static list of static size, in which case the static rule would detect it. The only case the dynamic rule would detect that the static rule won't is if the list itself --- not just the contents of the list --- is computed.

Do you have a case in mind that the static rule would not detect?

**Comment by katelynsills** [comment]
Do we have a timeline for a static check? Previously we've talked about static checks but we haven't written any, so I'm thinking that in the timespan of Zoe Alpha, there is no such static check. That's one of the reasons why I'd like to throw an error. I think the short term choice is between throwing an error and silently passing. Given that we are focusing on user comprehension and learning, I think throwing an error is better than silently passing.

**Comment by erights** [comment]
I suspected something like that. That's fine. But we should always have a clear longer term plan that actually makes sense, and always state in the interim kludge that it will be replaced with the long term plan.

**Comment by DavidBruant** [comment]
Another option i haven't seen listed could be that in a "development environment", the zcf in the contract is a wrapped version that performs additional checks (at runtime). The wrapping does not exist in the production version

This way, it's possible to perform checks (even potentially expensive ones) for when the dev designs/writes/tests the contract and the experience is better. And when the contract goes to prod, there are no checks.
The wrapping can be written as an option layer on top of the actual implementation 

As far as a static check, i briefly looked at whether TypeScript was expressive enough to express "this argument is expected to be an array of at least 2 elements" but i haven't found a solution yet.

**Comment by katelynsills** [comment]
Closed by #1009
